### PR TITLE
fix(workflows): preserve loop-group raw output on resume

### LINKED ISSUE #2585: test(workflows): pin the loop_group completion half of the when: whole-output rejection message
## Problem

The load error #2579 added for a `when:` that compares a `loop_group:` node's whole output to a
literal makes a **two-part** factual claim:

> A `loop_group:` node's output is the last iteration's raw text — unlike a `prompt:` node,
> declaring `output_format` does not replace it with the JSON document

- **(a)** `output_format` survives parse on a `loop_group:` (unlike on a `loop:`, where the
  transform drops it).
- **(b)** the group's completion nonetheless returns raw text, so surviving parse changes nothing.

#2579 added a schema-level test pinning **(a)** only
(`schemas.test.ts` → `dagNodeSchema — output_format survival by node type`).

**(b) is pinned by nothing.** If a future change taught the loop_group completion path to
synthesize `structuredOutput` from `node.output_format` — the way the prompt/command path already
does — then (b) goes false, the shipped error message starts misleading authors, and the schema
test keeps passing because it never touches `dag-executor.ts`.

Note the asymmetry: the `schema-capable` message's runtime half **is** independently pinned, by an
existing assertion in `dag-executor.test.ts`. Only the loop_group half is unguarded.

## Current state (verified, holds today)

`lastIterationStructuredOutput` exists only in `executeLoopNode`. The loop_group executor has no
such variable, and every terminal return there is `output: lastIterationOutput` bare.
`declaredFields` is set on the prompt/command path, the `workflow:` path and the resume
prepopulation — never in the loop_group completion path.

## Proposed Solution

One companion assertion in `dag-executor.test.ts`, beside the existing loop_group completion tests:
a `loop_group` with `output_format` set still produces plain-text output, with no
`structuredOutput` and no `declaredFields` on its `NodeOutput`.

**Best attached to whatever next changes loop_group completion shape** rather than landed on its
own — #2563 (loop structured completion) is the likely owner, since it is the change that would
make (b) false. If it lands there, the error message, the authoring guide's `when:` caution and the
constitution's case-law row all need revisiting in the same commit.

## Why this is worth tracking

The underlying claim was written three times across #2579 and was wrong or imprecise twice. Every
catch was a human reading prose against code; nothing in the repository held the claim. (a) is now
held. (b) is not.

## Scope

- Package(s): `workflows` (`dag-executor.test.ts` only — no production change)
- Breaking change? `No`
- Database changes needed? `No`

## Related

- #2579 (added the message and the (a)-half test)
- #2563 (loop structured completion — the change that would falsify (b))
- #2566 (the rejection this message belongs to)

### LINKED ISSUE #2587: fix(workflows): a loop_group's $grp.output.field resolves differently on a fresh run than after a resume
## Problem

A `loop_group` can currently resolve `$group.output.<field>` differently on a fresh run and after resume, even when the stored output bytes are identical.

The mismatch occurs when the group declares `output_format`. Fresh completion returns the group's raw final-iteration output without declared structured fields, while resume reconstruction can derive declared fields from the unsupported declaration. The same reference can therefore succeed on one path and fail on the other.

## Current product contract

A `loop_group` has a raw output boundary: its output is the final body output from the completing iteration.

`output_format` is not a structured-output contract for the group itself. Structured judgment belongs on a body node, and the group completion condition consumes that body result. An unsupported group-level declaration must not silently acquire meaning only during resume.

## Product intent

Fresh execution and resumed execution must expose exactly the same loop-group output contract.

- Byte-identical stored group output resolves identically before and after resume.
- Resume reconstruction cannot invent declared fields that fresh execution did not produce.
- The group boundary remains raw output; this issue does not add a new structured-output channel to `loop_group`.
- Unsupported configuration is handled consistently and visibly rather than partially affecting one execution path.

## Acceptance

- Whole-output and field references against a loop-group result behave identically on fresh and resumed runs.
- A group-level `output_format` cannot affect only the resumed path.
- Tests cover both a present undeclared JSON field and an absent declared field using byte-identical fresh/resumed output.
- Documentation, validation messages, and runtime behavior describe the same raw-output contract.

## Related

- #2563 - loop completion channels, delivered
- #2585 - companion regression-test gap
- #2456 and #2091 - strict field references and resume invariance

### PR DESCRIPTION
**Pull request by Wirasm**
## Problem and outcome

A completed `loop_group` exposes its final body output as raw text, but resume reconstruction was attaching the group's ignored `output_format` as a declared-field contract. The same stored JSON bytes therefore resolved fields differently depending on whether the downstream node ran fresh or after resume.

- **Outcome:** Byte-identical loop-group output now resolves identically fresh and resumed, including present undeclared fields and absent fields named only by the unsupported group schema.
- **Invariant:** A `loop_group` keeps one raw output boundary; resume must not invent a structured-output channel that fresh execution never produced.
- **Scope boundary:** Schema-capable prompt, command, loop, and workflow producers retain their existing strict resume reconstruction. Persistence, schemas, and YAML syntax are unchanged.
- **Root cause:** Generic resume prepopulation derived declared fields from every loaded node's `output_format`, including `loop_group`, even though fresh group completion ignores that field.

## Review guidance

- **Feedback requested:** Correctness of the raw-producer exception and fresh/resume regression coverage.
- **Start here:** `packages/workflows/src/dag-executor.ts:8838` — the typed discriminator keeps only loop-group resume output schemaless.
- **Review order:** Resume reconstruction, paired executor regressions, then the loader warning assertion.
- **Lower-attention areas:** Test fixture setup; it uses existing executor and discovery paths without production abstractions.
- **Known risk or uncertainty:** None within the stable workflow definition. Immutable workflow source across resume is separately designed in #2646.

## Solution

Resume prepopulation now uses the existing `isLoopGroupNode` discriminator to skip declared-field derivation for groups while preserving #2091 behavior for every schema-capable producer. Paired executor tests run the same raw JSON workflow fresh and from a byte-identical prior-success snapshot: an undeclared present field resolves on both paths, while a declared-but-absent field fails with the same missing-key error on both paths. A companion structured-loop resume test protects the neighboring schema-capable producer contract. Discovery coverage also pins the existing warning that group-level `output_format` is ignored.

## Behavior change

| | Before | After |
|---|---|---|
| **Present undeclared JSON field** | Resolved fresh; rejected after resume as outside the unsupported schema | Resolves to the same value fresh and resumed |
| **Absent field declared only on the group** | Failed fresh; resolved empty after resume | Fails with the same missing-key error fresh and resumed |

## Validation

- Pre-fix focused regression — both acceptance tests failed in the opposite fresh/resume directions described above.
- `cd packages/workflows && bun test src/dag-executor.test.ts` — 530 passed, 0 failed, 1,391 assertions.
- `cd packages/workflows && bun test src/loader.test.ts` — 284 passed, 0 failed, 762 assertions.
- `bun --filter @archon/workflows type-check` — passed.
- `bun run validate` — passed generated checks, all package type-checks, zero-warning lint, formatting, installer tests, package-isolated tests, and script tests.
- **Not verified:** Nothing material within #2587's stable-definition contract; workflow-source immutability across resume is tracked by #2646.

## Links

- Closes #2587
- Closes #2585
- Related #2563
- Related #2646
- Plan: https://github.com/coleam00/Archon/issues/2587#issuecomment-5352191988

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed workflow resume behavior so loop-group outputs remain consistent with fresh executions.
  * Preserved available undeclared output fields during resumed runs and consistently handled fields that are missing.
  * Ensured optional structured outputs resolve consistently when values are unavailable.
  * Ignored unsupported output formatting for loop groups while allowing valid workflow configurations to load successfully.

### CODE REVIEW

**Comment by Wirasm** [comment]
---
pr: 2649
base: dev
head: fix/2587-loop-group-resume-output
reviewed: 2026-08-20T08:43:12Z
reviewed_head: c1bed8bcd7f3aefbd381d0c139e9aabbd0eeaf37
verdict: READY TO MERGE
open_findings: 0
scopes: [code, seams, tests]
publication: https://github.com/coleam00/Archon/pull/2649#issuecomment-5352358531
---



# PR Review: #2649 — fix(workflows): preserve loop-group raw output on resume

## Signal

The correction is ready to merge at `c1bed8bcd7f3aefbd381d0c139e9aabbd0eeaf37`. Correction-focused review of `8e7709eacce617fa3a897b522cf2f09cac6bf8f4..c1bed8bcd7f3aefbd381d0c139e9aabbd0eeaf37` found that the new structured plain-loop resume test directly closes `R2` without changing production behavior. The [CodeRabbit Major](https://github.com/coleam00/Archon/pull/2649#discussion_r3819266854) supplied a concrete instance of existing `R1`: raw `loop_group` bytes can be reinterpreted after the same node ID becomes a structured `loop`. Independent reproduction proved the scenario, confirmed it predates this PR, and recorded it under [#2646](https://github.com/coleam00/Archon/issues/2646#issuecomment-5353524593), whose required invariant is that resume uses the same workflow bytes it started with. It is not a blocker for #2587's stable-definition parity contract. No correction-caused findings remain, and all local and exact-head GitHub validation passed.

- **Blocking:** 0
- **Non-blocking:** 0
- **Resolved since previous review:** 1 (`R2`)
- **Tracked follow-ups:** 1 (`R1` → [#2646 evidence](https://github.com/coleam00/Archon/issues/2646#issuecomment-5353524593))
- **Validation:** Green locally and in GitHub Actions at `c1bed8bcd7f3aefbd381d0c139e9aabbd0eeaf37`

## Findings

| ID | Severity | Finding | State |
|---|---|---|---|
| `R1` | Important | Resume rebuilds the historical output contract from mutable current workflow source | TRACKED FOLLOW-UP |
| `R2` | Important | Structured plain-loop resume semantics lack regression protection | FIXED |

## Detailed Findings

<details>
<summary><code>R1</code> — Keep workflow source immutable across resume</summary>

**Impact:** If a run resumes against different workflow source, a completed node's stored bytes can be paired with a different node kind, graph, or output declaration. In the verified case, `iterate` completes as a raw `loop_group` with `{"done":true}`, then the same ID becomes a structured `loop` with optional `note`. The original raw contract fails `missing-key`; current-definition hydration attaches the new declared fields and resolves `note` as empty. Output-contract persistence alone would still leave stale completed-node identity, changed graph semantics, and resource ownership unresolved.

**Evidence:** The [CodeRabbit thread](https://github.com/coleam00/Archon/pull/2649#discussion_r3819266854) was independently reproduced. `packages/workflows/src/dag-executor.ts:4010-4062` persists and returns raw group bytes; `packages/core/src/db/workflow-events.ts:250-267` restores only `Map<string, string>`; `packages/workflows/src/executor.ts:575-610` forwards that text-only snapshot; and `packages/workflows/src/dag-executor.ts:8832-8850` derives the contract from the currently loaded same-ID node. `packages/workflows/src/output-ref.ts:196-218` then maps a declared optional missing field to empty, while `packages/workflows/src/output-ref.ts:232-239` makes the original raw contract fail. The base implementation already reconstructed contracts from current source, so #2649 did not introduce the path. [#2646](https://github.com/coleam00/Archon/issues/2646#issuecomment-5353524593) explicitly requires that “a resumed run uses the same workflow bytes it started with” and now records this exact scenario and the legacy-event limitation.

**Required outcome:** Resolve [#2646](https://github.com/coleam00/Archon/issues/2646#issuecomment-5353524593) with a source model that guarantees a resumed run uses the exact workflow bytes and owned resources captured at run start, while new runs see later authoring edits. Legacy events contain no original producer contract, so that work must define an explicit compatibility fallback instead of inferring unavailable history.

**Found by:** `seam-analyzer`, `coderabbitai`; disposition independently verified by `code-reviewer`, `seam-analyzer`, and the focused persistence/hydration audit.

**Disposition:** TRACKED FOLLOW-UP — [the verified #2646 record](https://github.com/coleam00/Archon/issues/2646#issuecomment-5353524593) now captures the raw-group → structured-loop case, causal boundary, and legacy limitation. #2646 fully owns the mutable-source/output-identity seam. PR #2649 correctly restores raw `loop_group` parity under #2587's stable-definition scope, so the broader immutable-source design does not block it.

</details>

<details>
<summary><code>R2</code> — Structured plain-loop resume behavior is protected</summary>

**Impact:** Without a focused regression, a later refactor could broaden the raw-output exception from `loop_group` to both loop kinds while the raw-group and prompt-producer tests remained green, causing resumed structured loops to lose their declared-field contract.

**Evidence:** The correction adds `keeps a structured loop declared-field contract on resume` at `packages/workflows/src/dag-executor.test.ts:5340`. It defines a genuine plain `loop` with required `done` and optional `note`, preloads `{"done":true}` through `priorCompletedNodes`, proves exactly one downstream consumer executes, and asserts `note=[]`. If `packages/workflows/src/dag-executor.ts:8842-8845` excluded plain loops from `declaredFields` reconstruction, schemaless resolution would reject the missing key and both assertions would fail.

**Required outcome:** Add a focused DAG-resume test for a plain structured loop and prove a declared optional absent field retains its existing empty-value behavior.

**Found by:** `pr-test-analyzer`; correction independently verified by `code-reviewer` and `pr-test-analyzer`.

**Disposition:** FIXED at `c1bed8bcd7f3aefbd381d0c139e9aabbd0eeaf37` — the requested regression is present, discriminating, and green.

</details>

## Agent Coverage

| Scope | Result |
|---|---|
| code | `R1` tracked in #2646; `R2` fixed; no correction-caused findings |
| seams | `R1` tracked in #2646; no correction-caused seam findings |
| tests | `R2` fixed; the cross-definition regression belongs with #2646's immutable-source outcome |

Docs were not repeated because the correction changes only a test and the previous docs review had no finding.

## Validation

| Command | Result | Evidence |
|---|---|---|
| `bun test packages/workflows/src/dag-executor.test.ts -t 'raw loop_group|absent loop_group|structured loop declared-field'` | PASS | 3 tests passed, 16 assertions; raw-group parity and neighboring structured-loop resume semantics are protected together. |
| `bun test packages/workflows/src/dag-executor.test.ts` | PASS | 530 tests passed, 1,391 assertions. |
| `bun run validate` | PASS | Generated-source checks, all package type checks, zero-warning ESLint, Prettier, installer tests, workspace-isolated tests, and root script tests exited 0. |
| `git diff --check 8e7709eacce617fa3a897b522cf2f09cac6bf8f4 c1bed8bcd7f3aefbd381d0c139e9aabbd0eeaf37` | PASS | The bounded correction is one clean 57-line test-only addition; the worktree remained clean. |
| `gh issue view 2646 --json body,state,url` plus [published evidence](https://github.com/coleam00/Archon/issues/2646#issuecomment-5353524593) | PASS | Open issue #2646 owns same-workflow-bytes resume and now records the verified CodeRabbit scenario and legacy limitation. |
| Paginated GraphQL `reviewThreads` audit | PASS | 1 total thread, 1 resolved, 0 unresolved after the [evidence-backed reply](https://github.com/coleam00/Archon/pull/2649#discussion_r3819973670). |
| `gh pr view 2649 --json headRefOid,statusCheckRollup,mergeable,mergeStateStatus` | PASS | Head remained `c1bed8bcd7f3aefbd381d0c139e9aabbd0eeaf37`; Ubuntu, Windows, schema-upgrade, Docker, and CodeRabbit succeeded; the PR is open, non-draft, mergeable, and clean. |

## Verdict

**READY TO MERGE**

`R2` is fixed. `R1` is independently verified as a distinct tracked design issue in #2646, and no open Important or Critical finding remains in PR #2649.

File: packages/workflows/src/dag-executor.test.ts [hunk 1]
```diff
@@ -5202,6 +5280,120 @@ describe('executeDagWorkflow -- resume with priorCompletedNodes', () => {
...
+      undefined,
+      join(testDir, 'artifacts'),
+      join(testDir, 'state'),
+      join(testDir, 'logs'),
+      'main',
+      'docs/',
+      minimalConfig,
+      undefined,
+      undefined,
+      new Map([['iterate', '{"done":true}']])
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/workflows/src/dag-executor.test.ts:5390]: Verified: this same-ID definition-drift scenario is real. A raw `loop_group` completion persists only `node_output`; `getDagResumeSnapshot` reduces it to `Map<string, string>`; resume then derives `declaredFields` from the currently loaded node. If that ID is now a structured `loop`, missing optional `note` resolves empty instead of retaining the original raw `missing-key` behavior.

This path predates PR #2649. The PR restores fresh/resume parity for a stable `loop_group` definition; it does not introduce mutable-source resume. Persisting one output-contract tag here would address this value symptom but leave graph identity, node kind, workflow resources, and other source-derived behavior mutable. That heavier invariant is already owned by #2646: resume must use the exact workflow bytes captured at run start.

The reproduced case, causal boundary, and unavoidable legacy-event limitation are now recorded at https://github.com/coleam00/Archon/issues/2646#issuecomment-5353524593. The canonical PRP review records this Major under existing `R1` as `TRACKED FOLLOW-UP`: https://github.com/coleam00/Archon/pull/2649#issuecomment-5352358531.
